calliope/core/util/observed_dict.py

```diff
@@ -0,0 +1,73 @@
+
+"""
+Copyright (C) 2013-2019 Calliope contributors listed in AUTHORS.
+Licensed under the Apache 2.0 License (see LICENSE file).
+
+"""
+
+from collections import Mapping
+from calliope.core.attrdict import AttrDict
+
+
+class ObservedDict(dict):
+    """
+    Dictionary subclass which updates an observer when there is a change in the
+    values assigned to keys in the Dictionary.
+    """
+    def __init__(self, initial_dict, on_changed=None):
+        if isinstance(initial_dict, str):
+            initial_dict = AttrDict.from_yaml_string(initial_dict).as_dict()
+
+        super().__init__(initial_dict)
+
+        self.on_changed = on_changed
+
+        for k, v in initial_dict.items():
+            if isinstance(v, dict):
+                super().__setitem__(
+                    k, ObservedDict(v, on_changed=self.notify))
+        self.notify()
+
+    def __setitem__(self, key, value):
+        if isinstance(value, dict):
+            value = ObservedDict(value, on_changed=self.notify)
+        super().__setitem__(key, value)
+        self.notify()
+
+    def update(self, other=None, **kwargs):
+        if other is not None:
+            for k, v in other.items() if isinstance(other, Mapping) else other:
+                self[k] = v
+        for k, v in kwargs.items():
+            self[k] = v
+
+    def notify(self, updated=None):
+        if self.on_changed is not None:
+            return self.on_changed(self)
+
+
+class UpdateObserver(ObservedDict):
```

Comment:
I find the class name a little bit confusing, because it is a subclass of ObservedDict, which is a subclass of Dict, but the name doesn't have dict in it. Wounldn't `UpdateObserverDict` make more sense?

Reply:
yeah, sounds good